compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/KernelDispatch.cpp

```diff
@@ -29,6 +28,152 @@ static LogicalResult setRootConfigForPadPipeline(func::FuncOp entryPointFn,
       IREE::Codegen::DispatchLoweringPassPipeline::None);
 }
 
+// This function is based on the following table pulled from the AIEVec_MatMulOp
+// documentation in mlir-aie/include/aie/Dialect/AIEVec/IR/AIEVecOps.td
+//
+//   lhs                | rhs                | Accumulator
+//  :------------------:|:------------------:|:-----------------:
+//   `vector<4x16xi8>`  | `vector<16x8xi4>`  | `vector<4x8xi32>`
+//   `vector<4x8xi8>`   | `vector<8x8xi8>`   | `vector<4x8xi32>`
+//   `vector<4x4xi16>`  | `vector<4x8xi8>`   | `vector<4x8xi32>`
+//   `vector<4x2xi16>`  | `vector<2x8xi16>`  | `vector<4x8xi32>`
+//   `vector<2x8xi16>`  | `vector<8x8xi8>`   | `vector<2x8xi64>`
+//   `vector<4x8xi16>`  | `vector<8x4xi8>`   | `vector<4x4xi64>`
+//   `vector<2x4xi16>`  | `vector<4x8xi16>`  | `vector<2x8xi64>`
+//   `vector<4x4xi16>`  | `vector<4x4xi16>`  | `vector<4x4xi64>`
+//   `vector<4x2xi32>`  | `vector<2x4xi16>`  | `vector<4x4xi64>`
+//   `vector<4x8xbf16>` | `vector<8x4xbf16>` | `vector<4x4xf32>`
+//
+// An instruction size (m, n, k) is returned for each of the above cases. Cases
+// which are not covered by the table return a failure.
+//
+// Example: consider the first line of the table:
+//   `vector<4x16xi8>`  | `vector<16x8xi4>`  | `vector<4x8xi32>`
+//
+// This first line says that if the LHS is a i8 tensor, RHS is a i4 tensor and
+// the accumulator is a i32 tensor, then we should target a matmul instruction
+// with m = 4, n = 8, k = 16.
+static FailureOr<std::array<uint32_t, 3>> getAIEMatmulInstructionSize(
+    Type elTypeLhs, Type elTypeRhs, Type elTypeAcc) {
+  auto nBitsLhs = elTypeLhs.getIntOrFloatBitWidth();
+  auto nBitsRhs = elTypeRhs.getIntOrFloatBitWidth();
+  auto nBitsAcc = elTypeAcc.getIntOrFloatBitWidth();
+
+  bool allFloatingPoint = elTypeLhs.isa<FloatType>() &&
+                          elTypeRhs.isa<FloatType>() &&
+                          elTypeAcc.isa<FloatType>();
+
+  bool allInteger = elTypeLhs.isa<IntegerType>() &&
+                    elTypeRhs.isa<IntegerType>() &&
+                    elTypeAcc.isa<IntegerType>();
+
+  if (!allInteger && !allFloatingPoint) {
+    return failure();
+  }
+
+  if (allFloatingPoint) {
+    if (nBitsLhs == 16 && nBitsRhs == 16 && nBitsAcc == 32) {
+      return std::array<uint32_t, 3>{4, 4, 8};
+    }
+    // There is only 1 floating point case in the table.
+    return failure();
+  }
+
+  assert(allInteger);
+
+  // pack the three bitwidths into a single 32-bit unsigned integer. This
+  // facilitates the use of a switch statement below.
+  auto getTriplet = [](uint32_t a, uint32_t b, uint32_t c) -> uint32_t {
+    return a + (b << 8) + (c << 16);
+  };
+
+  // Sanity check: getTriplet values are computed at compile time.
+  static_assert(getTriplet(1, 2, 3) == 1 + 2 * 256 + 3 * 65536);
+
+  switch (getTriplet(nBitsLhs, nBitsRhs, nBitsAcc)) {
```

Comment:
Seems like a rather complicated way of just representing a table. I think you can just define a 
```
llvm::DenseMap<std::array<int, 3>, std::array<int, 3>> = {
 {8, 4, 32}, {4, 8, 16},
 ...
};
```

Reply:
This doesn't work, they key must have an llvm tombstone (more complex to extend with tombstone IMO than current design). I could use a std::map. Let me know if you'd like std::map instead of the change I've pushed now, happy to change again.